### PR TITLE
bpo-4630: Add cursor no-blink option for IDLE

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Add an option to toggle IDLE's cursor blink.

Co-Authored-By: Guilherme Polo <ggpolo@gmail.com>
Co-Authored-By: Roger Serwy <roger.serwy@gmail.com>

https://bugs.python.org/issue4630

### CODE REVIEW

File: Lib/idlelib/configdialog.py [hunk 1]
```diff
@@ -1976,6 +1982,13 @@ def create_page_general(self):
                 validatecommand=self.digits_only, validate='key',
         )
 
+        frame_cursor_blink_default = Frame(frame_editor, borderwidth=0)
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]:
```suggestion
        frame_cursor_blink = Frame(frame_window, borderwidth=0)
```

File: Lib/idlelib/configdialog.py [hunk 2]
```diff
@@ -1844,6 +1845,9 @@ def create_page_general(self):
                 frame_context: Frame
                     context_title: Label
                     (*)context_int: Entry - context_lines
+                frame_cursor_blink_default: Frame
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: 1. Remove 'default' from comment lines and all code lines, including pack and tracer lines.
The line_number stuff is not an example to follow (I will change it also in a separate PR).

2. Move comment lines to frame_window, as this applies to Shell, Editor, and Output windows.
I want to try putting this just after win_size.

3. Move code similarly.

File: Lib/idlelib/editor.py [hunk 3]
```diff
@@ -803,6 +807,16 @@ def colorize_syntax_error(self, text, pos):
             text.mark_set("insert", pos + "+1c")
         text.see(pos)
 
+    def UpdateCursorBlink(self):
+        "Update the cursor blink configuration."
+        cursorblink = idleConf.GetOption(
+                'main', 'EditorWindow', 'cursor-blink', type='bool')
+        if not cursorblink:
+            self.text.config(insertofftime=0)
```
Comment (terryjreedy) [inline Lib/idlelib/editor.py]:
```suggestion
            self.text['insertofftime'] = 0
```

File: Lib/idlelib/editor.py [hunk 4]
```diff
@@ -134,6 +134,10 @@ def __init__(self, flist=None, filename=None, key=None, root=None):
                         'main', 'EditorWindow', 'height', type='int'),
                 }
         self.text = text = MultiCallCreator(Text)(text_frame, **text_options)
+        # Store the current value of the insertofftime now so we can restore
+        # it if needed.
+        self.text._insertofftime = self.text['insertofftime']
```
Comment (terryjreedy) [inline Lib/idlelib/editor.py]: The attribute should be attached to self, not the tk text, as with other instance attributes.
```suggestion
        self.blink_off_time = self.text['insertofftime']
```
Corresponding change is suggested below.

The NMT tkinter 8.5 manual says that this is 300, but the tcl/tk 8.6 reference does not give a default, so grabbing the initial time seems safest for matching blinks on a particular system.

This set of new lines belongs further down the file, perhaps following 
`        self.set_notabs_indentwidth()`

File: Lib/idlelib/editor.py [hunk 5]
```diff
@@ -803,6 +807,16 @@ def colorize_syntax_error(self, text, pos):
...
 
+    def UpdateCursorBlink(self):
+        "Update the cursor blink configuration."
+        cursorblink = idleConf.GetOption(
+                'main', 'EditorWindow', 'cursor-blink', type='bool')
+        if not cursorblink:
+            self.text.config(insertofftime=0)
+        else:
+            # Restore the original value
+            self.text.config(insertofftime=self.text._insertofftime)
```
Comment (terryjreedy) [inline Lib/idlelib/editor.py]:
```suggestion
            self.text['insertofftime'] = self.blink_off_time
```
I prefer the pythonic option getting/setting.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Thank you *very* much for converting this overdue, overlooked patch.  I realize that adjusting the 9-year-old code to the highly revised configdialog was very much *not* 'cut and paste'.

AFAICT, the code works and changes properly persist between sessions.

After the revisions, a new test should be added to test_configdialog.  I expect to write but you can try if you want.

I want this merged in November for the December releases, so ping if I let this go more than a week after revision.

**Comment by ZackerySpytz** [comment]
Thank you, @terryjreedy, for the thorough review. I believe I have addressed all of
your comments.

I was unsure if the `cursor-blink` config option in `config-main.def` should be moved.

I have made the requested changes; please review again.

**Comment by ZackerySpytz** [comment]
@terryjreedy Ping.

**Review by terryjreedy — APPROVED** [review]
As I noted on the issue, the original news item was not true, but the reduced claim is.  Storing insertofftime on idleConf will make a 2nd patch much easier.

The new update_cursor_blink was tested manually, though a unittest should be added.

**Pull request merged**